Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37249 Consolidate usage on torch::jit::toPyObject in RPC request_callback
* #36811 Add using torch::utils::Future to simplify code in RRefContext
* **#36805 Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks**
* #36785 Prevent RRef unpickle to block waiting for OwnerRRef creation

Before this commit, RRef.to_here() will block one RPC thread on the owner
until the OwnerRRef is created. This leads to two problems:

1. The correctness of RRef protocol depends on the number of threads
   in the RPC agent implementation. If there are not enough thread,
   RRef protocal itself can deadlock.
2. It unnecessarily occupies a thread on the owner.

This commit addresses the problem by letting RRefContext return a
Future if the OwnerRRef has not been created. Then the request processing
logic for RRef.to_here() can add callbacks to the Future to resume
when the OwnerRRef is created.

Differential Revision: [D21093847](https://our.internmc.facebook.com/intern/diff/D21093847)

Commit:
Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks

[ghstack-poisoned]

mrshenli committed:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"




[ghstack-poisoned]

Commit:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"




[ghstack-poisoned]

mrshenli committed:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"




[ghstack-poisoned]

mrshenli committed:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"




[ghstack-poisoned]

mrshenli committed:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"


Differential Revision: [D21093847](https://our.internmc.facebook.com/intern/diff/D21093847)

[ghstack-poisoned]

mrshenli committed:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"


Differential Revision: [D21093847](https://our.internmc.facebook.com/intern/diff/D21093847)

[ghstack-poisoned]

mrshenli committed:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"


Differential Revision: [D21093847](https://our.internmc.facebook.com/intern/diff/D21093847)

[ghstack-poisoned]

rohan-varma commented:
Is the need for this to enable training with `num_send_recv_threads=1`? Would be good to clarify this and the expected before/after behavior in the PR description.

rohan-varma commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -301,16 +307,26 @@ c10::intrusive_ptr<OwnerRRef> RRefContext::createOwnerRRef(
   } else {
     // Scenario (2) retrieving an existing RRef
-    return c10::static_intrusive_pointer_cast<OwnerRRef>(iter->second);
+    return std::make_shared<
```
I guess this implicitly assumes that the future API marks a future as completed if it's created with a value `T`, although it doesn't seem like this holds true for the ivalue generic future. Not sure if this assumption might introduce bugs when we attempt to unify the futures.

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -312,13 +312,24 @@ void RequestCallbackImpl::processRpc(
+
+      if (futureOwner->completed()) { // optional fast-path
+        // the OwnerRRef has been created
+        auto& rref = futureOwner->wait();
```
Can we use `constValue()` since we are sure the value is valid? Might be a bit confusing to readers that this wait isn't actually a blocking wait

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -312,13 +312,24 @@ void RequestCallbackImpl::processRpc(
+        }
+      }
+
+      futureOwner->addCallback([responseFuture, messageId, futureOwner]() {
```
This can now be run in 2 different situations right? When the owner hasn't confirmed the RRef, or when the rref value hasn't been set. Does there need to be any difference in behavior for these 2 situations?

mrshenli replied:
> When the owner hasn't confirmed the RRef, or when the rref value hasn't been set. 

This is on owner. So it is actually when 1) the OwnerRRef hasn't been created by the `rpc.remote` message, 2) when the rref value hasn't been set

> This can now be run in 2 different situations right?

This cb is launched when `OwnerRRef` is created. And the nested cb on `whenValueSet` when be launched when the value is ready.

> Does there need to be any difference in behavior for these 2 situations?

Not to my knowledge. They are different cbs. So should be fine?

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -333,53 +344,65 @@ void RequestCallbackImpl::processRpc(
+        auto& rref = futureOwner->wait();
+        if (rref->hasValue()) {
+          auto value = rref->getValue();
+          py::object pyValue;
```
Nit: Should we move this python conversion/serialization to someplace like rpc/utils.cpp?

mrshenli replied:
Sure, will do that in a followup PR.

mrshenli replied:
added in #37249

rohan-varma commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -253,7 +254,12 @@ c10::intrusive_ptr<OwnerRRef> RRefContext::getOrCreateOwnerRRef(
     owners_[rref->rrefId()] = rref;
-    ownerCV_.notify_all();
+    // ownerCV_.notify_all();
+    const auto pendingOwnerIter = pendingOwners_.find(rrefId);
```
Why does `getOrCreateOwnerRRef` not have to add to `pendingOwners_`  if the pendingOwner is not found, but `getOwnerRRef` does?

mrshenli replied:
Because the code here creates the OwnerRRef, while `getOwnerRRef` does not have enough info to create the `OwnerRRef` and have to wait for `getOrCreateOwnerRRef` to create it. This traces back to where these functions are called. The `getOrCreateOwnerRRef` is called in `SCRIPT_REMOTE_CALL` and `PYTHON_REMOTE_CALL` request processors, which always create the `OwnerRRef`

This is a good catch. We should have renamed the function to `createOwnerRRef` and let me try if we can get rid of the retrieving part. This was originally called `getOrCreateOwnerRRef` back when both `rpc.remote()` and `rref.to_here()` can create `OwnerRRefs`. I think we can simplify that logic now.

mrshenli replied:
Cannot get rid of the retrieving logic from `getOrCreateOwnerRRef`, because unpickle can also create the RRef. So the main difference between `getOrCreateOwnerRRef` and `getOwnerRRef` is whether we have the TypePtr information. As long as TypePtr is available, we can create and do not need to return a future.

mrshenli replied:
Let me add above comments to those functions.

rohan-varma replied:
Thanks for clarifying it!

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1397,6 +1397,62 @@ def test_single_threaded_rref_owner(self):
+
+            # makes sure there is no existing OwnerRRefs on dst
+            info = rpc.rpc_sync(dst, get_rref_debug_info)
+            self.assertEqual(0, int(info["num_owner_rrefs"]))
```
test looks fine to me, although, is there a way to check the behavior by ensuring that the remote end processes `rref.to_here()` before `rpc.remote()` 

I'm assuming the reason for this PR is that before with single threaded, the thread could block on a `to_here()`. But then it would never process the `rpc.remote()`, hence hanging forever.

mrshenli replied:
> I'm assuming the reason for this PR is that before with single threaded, the thread could block on a to_here(). But then it would never process the rpc.remote(), hence hanging forever.

Yes. `to_here` should not block, as it does not contain any user code.

>  is there a way to check the behavior by ensuring that the remote end processes rref.to_here() before rpc.remote()

Didn't figure out a way to test this with the current code. We need to somehow delay sending the first `rpc.remote` message and send the `to_here` message first. We might need a dedicated test RPC agent which allows application to control its send queue.

mrshenli commented:
> Is the need for this to enable training with num_send_recv_threads=1? Would be good to clarify this and the expected before/after behavior in the PR description.

Good point. This is intended to address #36755 for TensorPipe. Let me add that in PR description

Commit:
Update on "Prevent RRef.to_here() to block an RPC thread on the callee using Future callbacks"


Before this commit, RRef.to_here() will block one RPC thread on the owner
until the OwnerRRef is created. This leads to two problems:

1. The correctness of RRef protocol depends on the number of threads
   in the RPC agent implementation. If there are not enough thread,
   RRef protocal itself can deadlock.
2. It unnecessarily occupies a thread on the owner.

This commit addresses the problem by letting RRefContext return a
Future if the OwnerRRef has not been created. Then the request processing
logic for RRef.to_here() can add callbacks to the Future to resume
when the OwnerRRef is created.

Differential Revision: [D21093847](https://our.internmc.facebook.com/intern/diff/D21093847)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Looks good! Thanks for addressing the comments and clarifying
